### ISSUE TITLE
Data Download

### ISSUE THREAD

**Issue by mbabker** [joomla-projects/privacy-framework#3]
Various laws have a clause which states that users have the right to access their data.  We should provide a system that helps create this data download.

If we can create a fully automated system, that'd be best, otherwise we'll need to provide the tooling for site owners to generate this download.  There should also be some form of logging and notifications to site owners.

**Comment by Ruud68** [comment]
>For me a long shot really really nice to have goal here is a way to say "I'd like my data" and Joomla can build a ZIP package containing information specified by extensions reporting they have user data and what fields that data should include

+1 for this, as a site maintainer in the worst case scenario I see myself having a day time job on this. Automation is key.

This could also be handled by the plugin system discussed here: https://github.com/joomla-projects/privacy-framework/issues/1#issuecomment-386277992

**Comment by manojLondhe** [comment]
> For me a long shot really really nice to have goal here is a way to say "I'd like my data" and Joomla can build a ZIP package containing information specified by extensions reporting they have user data and what fields that data should include

+1, I was thinking on same lines. 

It's called as [Subject Access Request](https://ico.org.uk/for-organisations/guide-to-data-protection/principle-6-rights/subject-access-request/)

Here is what I have in mind.

**Flow 1:**

- User asks for his data (SAR)
- Add entry in database queue (status pending)
- Notify admin about pending entry
- Notify user about his request is pending
> For me a long shot really really nice to have goal here is a way to say "I'd like my data" and Joomla can build a ZIP package containing information specified by extensions reporting they have user data and what fields that data should include

+1, I was thinking on same lines. 

It's called as [Subject Access Request](https://ico.org.uk/for-organisations/guide-to-data-protection/principle-6-rights/subject-access-request/)

Here is what I have in mind.

**Flow 1:**

- User asks for his data (SAR)
- Add entry in database queue (status pending)
- Notify admin about pending entry
- Notify user about his request is pending

**Flow 2:**

Let's say there is a task / cron job that processes such requests

- Cron checks for all pending requests
- Cron might have batching (let's say 10 requests per run)
- Cron triggers plugins to get data
- Plugins return data in specified format (see flow 2.1)
- If any of the plugin has returned data, create a folder per request
- Create one or more CSV / JSON files based on data each plugin has returned
- Create a zip of this folder
- Update database entry (status: ready)
- Notify admin (not sure if this is needed)
- Notify user

**Flow 2.1:**
Let's say there are plugins of type `privacy` for every component (or extension) and they have a trigger `onPrepareSAR($user)`

A plugin for `com_users` might have code like this

```
public function onPrepareSAR($user)
{
    $data   = array();

    // Get data of user by running DB query
    $data['profile']['headers'] = // array of columns;
    $data['profile']['records']  = // array of records;

   return $data;   
}
```

A plugin for `com_content` might have code like this

```
public function onPrepareSAR($user)
{
    $data   = array();

    // Get data of user by running DB query
    $data['articles']['headers'] = // array of columns;
    $data['articles']['records']  = // array of records;

   return $data;   
}
```

A plugin for `com_kunena` might have code like this

```
public function onPrepareSAR($user)
{
    $data   = array();

    // Get data of user by running DB query
    $data['posts']['headers'] = // array of columns;
    $data['posts']['records']  = // array of records;

   return $data;   
}
```

A plugin for `com_myecommerceextension` might have code like this

```
public function onPrepareSAR($user)
{
    $data   = array();

    // Get data of user by running DB query
    $data['orders']['headers'] = // array of columns;
    $data['orders']['records']  = // array of records;

   return $data;   
}
```

and so on ...

Based on above data, cron job can create a CSV like profile.csv, articles.csv, posts.csv, orders.csv, which can be zipped later.

**Few more things**
- we need a view where user can request data (frontend)
- we need a view where admin can view pending requests (backend)

**Attached diagrams**
Flow 1
![sar-request-1](https://user-images.githubusercontent.com/1375006/39809203-c8186b46-539e-11e8-9eec-3cc086615187.jpg)

Flow 2
![sar-request-2](https://user-images.githubusercontent.com/1375006/39809372-375b4f32-539f-11e8-8896-6e231aa16fc7.jpg)

I hope his makes sense.

**Comment by mbabker** [comment]
Export workflow implemented via #53
